client/gutenberg/editor/async-loader.jsx

```diff
@@ -0,0 +1,60 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { Component } from 'react';
+
+export const asyncLoader = ( { promises, loading, success, failure } ) =>
+	class AsyncLoader extends Component {
+		state = {
+			results: Object.keys( promises ).reduce(
+				( output, key ) => ( { ...output, [ key ]: null } ),
+				{}
+			),
+		};
+
+		componentDidMount() {
+			// here we want to make sure all promises "resolve" even when
+			// they reject because Promise.all will return a single value
+			// on the first rejection instead of an array as when they all
+			// resolve. this is for our own accounting and ensures that
+			// we wait for all promises to fulfill
+			const runners = Object.keys( promises ).map( key =>
+				promises[ key ]
+					.then( a => {
+						this.setState( state => ( { results: { ...( state || {} ).results, [ key ]: a } } ) );
+
+						return a;
+					} )
+					.then( a => [ true, a ], a => [ false, a ] )
+			);
+
+			Promise.all( runners ).then( resolutions => {
+				const keys = Object.keys( promises );
+				const [ results, successful ] = resolutions.reduce(
+					( [ output, allSuccess ], [ wasSuccess, data ], index ) => [
+						{ ...output, [ keys[ index ] ]: data },
+						allSuccess && wasSuccess,
+					],
+					[ {}, true ]
+				);
+
+				this.setState( { results, successful } );
```

Comment:
Here we set the `results` state to a value it already has, because there are `setState` calls as individual promises resolve. This one just fills in the errors? Let's make up our minds and set the `results` state only at one place: when individual promise resolves, or when all resolve.